common/src/character_controller.rs

```diff
@@ -77,4 +89,66 @@ impl CharacterControllerPass<'_> {
             self.position.local = transition_xf * self.position.local;
         }
     }
+
+    /// Checks for collisions when a character moves with a character-relative displacement vector of `relative_displacement`.
+    fn check_collision(&self, relative_displacement: &na::Vector3<f32>) -> CollisionCheckingResult {
+        // Split relative_displacement into its norm and a unit vector
+        let relative_displacement = relative_displacement.to_homogeneous();
+        let displacement_sqr = relative_displacement.norm_squared();
+        if displacement_sqr < 1e-16 {
+            // Fallback for if the displacement vector isn't large enough to reliably be normalized.
+            // Any value that is sufficiently large compared to f32::MIN_POSITIVE should work as the cutoff.
+            return CollisionCheckingResult::stationary();
+        }
+
+        let displacement_norm = displacement_sqr.sqrt();
+        let displacement_normalized = relative_displacement / displacement_norm;
+
+        let ray = ray_tracing::Ray::new(math::origin(), displacement_normalized);
+
+        let ray_tracing_result =
+            ray_tracing::trace_ray(self.position.local, &ray, displacement_norm.tanh());
+
+        let allowed_displacement = math::translate(
+            &ray.position,
+            &math::lorentz_normalize(&ray.ray_point(ray_tracing_result.tanh_distance)),
+        );
+
+        CollisionCheckingResult {
+            allowed_displacement,
+            collision: ray_tracing_result
+                .intersection
+                .map(|intersection| Collision {
+                    // RayTracingResult has its `normal` given relative to the character's original position,
+                    // but we want the normal relative to the character after the character moves to meet the wall.
+                    normal: (math::mtranspose(&allowed_displacement) * intersection.normal)
+                        .xyz()
+                        .normalize(),
+                }),
+        }
+    }
+}
+
+struct CollisionCheckingResult {
+    /// Multiplying the character's position by this matrix will move the character as far as it can up to its intended
+    /// displacement until it hits the wall.
+    allowed_displacement: na::Matrix4<f32>,
+    collision: Option<Collision>,
+}
+
+struct Collision {
+    /// This collision normal faces away from the collision surface and is given in the perspective of the character
+    /// _after_ it is transformed by `allowed_displacement`.
+    normal: na::Vector3<f32>,
```

Comment:
Why is this a `Vector3` when `RayTracingIntersection`'s is a `Vector4`?

Reply:
The normal vector here is in relation to the character, and the implication is that the vector represents a direction starting at the character's center. Because of that, the `w` coordinate is always 0, so we enforce that invariant by making it a `Vector3` instead of a `Vector4`. Converting a `Vector4` to this `Vector3` effectively performs the necessary projection to ensure this invariant holds.